### PR TITLE
feat: two-step payments introduced

### PR DESCRIPTION
**Pull request by beniaminmunteanu**
This PR implements issue #295 

Includes FE

image

### CODE REVIEW

File: packages/wallet/backend/src/app.ts [hunk 1]
```diff
@@ -179,6 +179,11 @@ export class App {
 
     // outgoing payment routes
     router.post('/outgoing-payments', isAuth, outgoingPaymentController.create)
+    router.post(
```
Comment (dragosp1011) [inline packages/wallet/backend/src/app.ts:182]: I think a better separation of endpoints will be
1. `POST /quotes` - as at the first request a quote is created and not an outgoing payments
2. `POST /outgoing-payments` - same endpoint that exists now, just that will get the quote as a required param

File: packages/wallet/frontend/src/pages/account/[accountId].tsx [hunk 2]
```diff
@@ -60,7 +60,7 @@ const AccountPage: NextPageWithLayout<AccountPageProps> = ({
             }
             className="group flex aspect-square h-24 w-24 flex-col items-center justify-center -space-y-1 rounded-lg border border-green-5 bg-white shadow-md hover:border-green-6"
           >
-            <New className="h-7 w-7" />
+            <New className="h-9 w-7" />
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/pages/account/[accountId].tsx:63]: The icons vary in terms of their heights and widths.

File: packages/wallet/frontend/src/pages/transfer/send.tsx [hunk 3]
```diff
@@ -106,15 +126,26 @@ const SendPage: NextPageWithLayout<SendProps> = ({ accounts }) => {
...
+                      handleAcceptQuote(response.data?.id || '')
+                      closeDialog
+                    }}
+                    onClose={closeDialog}
+                  />
+                )
+              } else {
+                openDialog(
+                  <ErrorDialog
+                    content="Something went wrong while fetching your Qoute. Please try again"
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/pages/transfer/send.tsx:144]:
```suggestion
                    content="Something went wrong while fetching your quote. Please try again."
```

File: packages/wallet/frontend/src/pages/transfer/send.tsx [hunk 4]
```diff
@@ -106,15 +126,26 @@ const SendPage: NextPageWithLayout<SendProps> = ({ accounts }) => {
...
-                  redirectText="Go to your accounts"
-                />
-              )
+              if (response.data) {
+                openDialog(
+                  <QuoteDialog
+                    quote={response.data}
+                    paymentType={sendForm.getValues().paymentType}
+                    onAccept={() => {
+                      handleAcceptQuote(response.data?.id || '')
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/pages/transfer/send.tsx:135]: Why is this needed?

File: packages/wallet/backend/src/outgoingPayment/controller.ts [hunk 5]
```diff
@@ -25,14 +30,34 @@ export class OutgoingPaymentController implements IOutgoingPaymentController {
...
         receiver,
         description
       )
+      res.status(200).json({ success: true, message: 'SUCCESS', data: quote })
+    } catch (e) {
+      next(e)
+    }
+  }
+
+  acceptQuote = async (
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/outgoingPayment/controller.ts]: Should this be moved into the `quote` module?

File: packages/wallet/backend/src/outgoingPayment/service.ts [hunk 6]
```diff
@@ -62,34 +63,51 @@ export class OutgoingPaymentService implements IOutgoingPaymentService {
...
     const payment = await this.deps.rafikiClient.createOutgoingPayment(
       paymentPointerId,
-      quote.id,
+      quoteId,
       description
     )
 
     return Transaction.query().insert({
-      paymentPointerId: existingPaymentPointer.id,
+      paymentPointerId: paymentPointerId,
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/outgoingPayment/service.ts:35]:
```suggestion
      paymentPointerId,
```

File: packages/wallet/backend/src/outgoingPayment/service.ts [hunk 7]
```diff
@@ -34,7 +35,7 @@ export class OutgoingPaymentService implements IOutgoingPaymentService {
     isReceive: boolean,
     receiver: string,
     description?: string
-  ): Promise<Transaction> {
+  ): Promise<Quote> {
```
[the remark this answers is not in the document]
Comment (beniaminmunteanu) [reply packages/wallet/backend/src/outgoingPayment/service.ts:38]: what abbout the acceptQuote then, shouldn't it also be moved to the quote service?
TBH i first had a quote service, but then I realised that both of these would be a better fit in the outgoing payments service.

Or so you preffer having both the create outgoing payment and the accept quote inside of the quotes service?

File: packages/wallet/frontend/src/components/dialogs/QuoteDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                    <br />
+                    Receiver gets: {receiveAmount.amount}
+                    <br />
+                    Fees paid by{' '}
+                    {paymentType === PAYMENT_RECEIVE ? 'you' : 'receiver'}:{' '}
+                    {feesAmount.amount}
+                  </p>
+                  <div className="mt-5 flex w-full flex-col justify-between space-y-3 sm:flex-row-reverse sm:space-y-0">
+                    <Button
+                      aria-label="close dialog"
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/components/dialogs/QuoteDialog.tsx:76]:
```suggestion
                      aria-label="accept quote"
```

File: packages/wallet/backend/src/outgoingPayment/service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (beniaminmunteanu) [reply packages/wallet/backend/src/outgoingPayment/service.ts:38]: will come up with the other option soon

File: packages/wallet/frontend/src/utils/helpers.ts [hunk 9]
```diff
@@ -57,3 +57,15 @@ export const formatDate = (date: string): string => {
     minute: '2-digit'
   })
 }
+
+export const getFeesValue = (
+  send: FormatAmountArgs,
+  receive: FormatAmountArgs
+): FormattedAmount => {
+  const fees = Number(send.value) - Number(receive.value)
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/utils/helpers.ts:65]: We should not use `Number(/*...*/)`, since everything comes as a `bigint`. In a particular case, if `send.value` is greater than [`Number.MAX_SAFE_INTEGER`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) we might get some weird results.

File: packages/wallet/backend/src/outgoingPayment/outgoing-payment-service.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  async create(quoteId: string): Promise<Transaction> {
+    const quote = await this.deps.rafikiClient.getQuote(quoteId)
+
+    const paymentPointerId = quote.paymentPointerId
+
+    const incomingPayment =
+      await this.deps.incomingPaymentService.getPaymentDetailsByUrl(
+        quote.receiver
+      )
+    const description = incomingPayment.description
```
Comment (dragosp1011) [inline packages/wallet/backend/src/outgoingPayment/outgoing-payment-service.ts:26]: `const { description, assetCode, value } = incomingPayment`

File: packages/wallet/backend/src/outgoingPayment/service.ts [hunk 11]
```diff
@@ -62,34 +63,51 @@ export class OutgoingPaymentService implements IOutgoingPaymentService {
...
     )
 
     return Transaction.query().insert({
-      paymentPointerId: existingPaymentPointer.id,
+      paymentPointerId: paymentPointerId,
       paymentId: payment.id,
-      assetCode: asset.code,
-      value: BigInt(amount * 10 ** asset.scale),
+      assetCode,
+      value: value,
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/outgoingPayment/service.ts]:
```suggestion
      value,
```

File: packages/wallet/backend/src/outgoingPayment/service.ts [hunk 12]
```diff
@@ -1,124 +1,44 @@
...
     )
 
     return Transaction.query().insert({
-      paymentPointerId: existingPaymentPointer.id,
+      paymentPointerId: paymentPointerId,
       paymentId: payment.id,
-      assetCode: asset.code,
-      value: BigInt(amount * 10 ** asset.scale),
+      assetCode,
+      value: BigInt(value),
```
Comment (dragosp1011) [inline packages/wallet/backend/src/outgoingPayment/service.ts:38]: should it be `BigInt(amount * 10 ** asset.scale)`

**Review by raducristianpopa — CHANGES_REQUESTED** [review]

File: packages/wallet/backend/src/outgoingPayment/controller.ts [hunk 13]
```diff
@@ -20,19 +20,12 @@ export class OutgoingPaymentController implements IOutgoingPaymentController {
...
 
-      const transaction = await this.deps.outgoingPaymentService.create(
-        userId,
-        paymentPointerId,
-        amount,
-        isReceive,
-        receiver,
-        description
-      )
+      const transaction = await this.deps.outgoingPaymentService.create(quoteId)
```
Comment (rico191013) [inline packages/wallet/backend/src/outgoingPayment/controller.ts:27]: maybe createById is better?

File: packages/wallet/backend/src/outgoingPayment/service.ts (hunk 12 above)
Comment (beniaminmunteanu) [reply packages/wallet/backend/src/outgoingPayment/service.ts:38]: Hmm i got it from the incoming payment. But I guess i should get it from the sendAmount from the quote.

File: packages/wallet/frontend/src/pages/transfer/send.tsx (hunk 4 above)
Comment (beniaminmunteanu) [reply packages/wallet/frontend/src/pages/transfer/send.tsx:135]: @Tymmmy

File: packages/wallet/backend/src/outgoingPayment/service.ts [hunk 14]
```diff
@@ -1,124 +1,48 @@
...
-    description?: string
-  ): Promise<Transaction> {
-    const existingPaymentPointer = await PaymentPointer.query().findById(
-      paymentPointerId
-    )
-
-    if (!existingPaymentPointer || !existingPaymentPointer.active) {
-      throw new BadRequest('Invalid payment pointer')
-    }
+  async createById(quoteId: string): Promise<Transaction> {
```
Comment (dragosp1011) [inline packages/wallet/backend/src/outgoingPayment/service.ts:17]: `createByQuoteId`

**Review by dragosp1011 — APPROVED** [review]

**Review by dragosp1011 — APPROVED** [review]

File: packages/wallet/backend/src/quote/service.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    const asset = await this.deps.rafikiClient.getAssetById(assetId)
+    if (!asset) {
+      throw new NotFound()
+    }
+
+    const value = BigInt((params.amount * 10 ** asset.scale).toFixed())
+
+    let paymentUrl = params.receiver
+    if (!incomingPaymentRegexp.test(params.receiver)) {
+      paymentUrl = await this.createReceiver({
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/quote/service.ts:71]: `createReceiver` should be a part of the IncomingPayments.

File: packages/wallet/frontend/src/components/dialogs/QuoteDialog.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+                <div className="flex flex-col items-center justify-center px-4">
+                  <Send strokeWidth={2} className="h-16 w-16" />
+                  <p className="text-center font-semibold text-turqoise">
+                    You will be deducted: {sendAmount.amount}
+                    <br />
+                    Receiver gets: {receiveAmount.amount}
+                    <br />
+                    Fees paid by{' '}
+                    {paymentType === PAYMENT_RECEIVE ? 'you' : 'receiver'}:{' '}
+                    {fee.amount}
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/components/dialogs/QuoteDialog.tsx:72]: https://github.com/interledger/testnet/pull/409#discussion_r1231981659 👈

File: packages/wallet/frontend/src/pages/account/[accountId].tsx (hunk 2 above)
Comment (raducristianpopa) [inline packages/wallet/frontend/src/pages/account/[accountId].tsx:63]: https://github.com/interledger/testnet/pull/409#discussion_r1231984219 👈

File: packages/wallet/frontend/src/utils/helpers.ts [hunk 17]
```diff
@@ -57,3 +57,15 @@ export const formatDate = (date: string): string => {
...
+  send: FormatAmountArgs,
+  receive: FormatAmountArgs
+): FormattedAmount => {
+  const fees = BigInt(send.value) - BigInt(receive.value)
+  return formatAmount({
+    assetCode: send.assetCode,
+    assetScale: send.assetScale,
+    value: fees.toString()
+  })
+}
```
Comment (raducristianpopa) [inline packages/wallet/frontend/src/utils/helpers.ts:71]:
```suggestion
export const getFee = (
  send: FormatAmountArgs,
  receive: FormatAmountArgs
): FormattedAmount => {
  const fee = BigInt(send.value) - BigInt(receive.value)
  return formatAmount({
    assetCode: send.assetCode,
    assetScale: send.assetScale,
    value: fee.toString()
  })
}
```

File: packages/wallet/backend/src/incomingPayment/service.ts [hunk 18]
```diff
@@ -16,11 +18,7 @@ interface IIncomingPaymentService {
   ) => Promise<Transaction>
   getPaymentDetailsByUrl: (url: string) => Promise<PaymentDetails>
   createIncomingPaymentTransactions: (
-    paymentPointerId: string,
-    amount: bigint | null,
-    asset: Asset,
-    description?: string,
-    expiresAt?: string
+    params: CreateIncomingPaymentParams
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/incomingPayment/service.ts:21]: We should choose if we use `params` or `args`.

File: packages/wallet/frontend/src/pages/account/[accountId].tsx (hunk 2 above)
Comment (Tymmmy) [reply packages/wallet/frontend/src/pages/account/[accountId].tsx:63]: I somehow missed this, sorry :D

File: packages/wallet/backend/src/incomingPayment/service.ts (hunk 18 above)
Comment (beniaminmunteanu) [reply packages/wallet/backend/src/incomingPayment/service.ts:22]: for now we used Params the most, and I'd continue with params, since args is what the generated graphql uses

Comment (beniaminmunteanu) [reply packages/wallet/backend/src/incomingPayment/service.ts:22]: will change what I encounter as args to params aswell from now on

**Review by Tymmmy — APPROVED** [review]

**Review by dragosp1011 — APPROVED** [review]

**Review by raducristianpopa — APPROVED** [review]
